Minor bugfixes

nznaza opened the pull request:
Fix Mouse Capture. 
All mouse actions on the overlay when is not clickable fall thought the bottom application (when flicking the mouse over the overlay no longer takes focus)

Fixed Framerate limiter
Framerate limiting now correctly applies using the sleep ticks (10000 ticks in a ms, 10000000 ticks in a second) (https://learn.microsoft.com/en-us/dotnet/api/system.timespan.ticks?view=netframework-4.5.2), added more framerate options.

Added Host Opacity Option.
A big blob of a gray area can be distracting even if it can be buried by other windows, added an option to make it 25, 50, 75 and 100% (Off) opaque. However, if the application doesn't start at 100% opacity, Discord may not detect it, so always starts with 100% opacity.

Overlay Host now controls Overlay Form Size
Resizing the host automatically resizes the Overlay to the correct size

nznaza commented:
My git configuration modified whitespace in more lines than needed

flamewave000 commented:
You'll need to resubmit this using tabs and not spaces

flamewave000 commented on DirectXHost.cs:
```diff
@@ -8,230 +8,261 @@
+        Graphics gfxScreenshot;
+        private void RenderCallback()
+        {
+            if (_dxForm.WindowState != FormWindowState.Minimized)
```
What is the purpose of this logic inversion?

nznaza replied:
WindowState can be Maximized (i.e. user snaps the window, keyboard shortcut, etc) which FormWindowState.Normal would ignore

flamewave000 commented on DirectXHost.cs:
```diff
@@ -8,230 +8,261 @@
+                _overlayTarget.Invalidate();
+            }
+            // Calculate Frame Limiting
+            //long end = DateTime.Now.Ticks;
+            //long duration = end - time;
+            int framerate = Settings.frameRate;
+            if (Settings.frameRate > 0)
+                Thread.Sleep(new TimeSpan(10000000 / Settings.frameRate));
```
I agree the original was missing the frame limit reference, which was accidental. But the proposed change is not an accurate frame limit as it does not take into account the actual duration of the frame. Should be the following:
```C#
if (Settings.frameRate > 0) {
	long end = DateTime.Now.Ticks;
	long duration = end - time;
	int timeout = (1000 / Settings.frameRate) - (int)new TimeSpan(duration).TotalMilliseconds;
	if (timeout > 0)
		Thread.Sleep(timeout);
}
```

nznaza replied:
New proposed change uses Stopwatch which is more accurate.

flamewave000 commented on DirectXHost.cs:
```diff
@@ -8,230 +8,261 @@
+            {
+                DoubleBuffered = true;
+            }
+            public void SetFormTransparent()
+            {
+                IntPtr Handle = this.Handle;
+                PreviousStyle = GetWindowLong(Handle, GWL_EXSTYLE);
+                SetWindowLong(Handle, GWL_EXSTYLE, Convert.ToInt32(PreviousStyle | WS_EX_LAYERED | WS_EX_TRANSPARENT));
+            }
+
+            public void SetFormNormal()
+            {
+                IntPtr Handle = this.Handle;
+                SetWindowLong(Handle, GWL_EXSTYLE, Convert.ToInt32(PreviousStyle | WS_EX_LAYERED));
+            }
+        }
```
NIT: The local variables `IntPtr Handle` are unnecessary. The second function can be shortened to a simple expression.

```C#
public void SetFormTransparent()
{
	PreviousStyle = GetWindowLong(Handle, GWL_EXSTYLE);
	SetWindowLong(Handle, GWL_EXSTYLE, Convert.ToInt32(PreviousStyle | WS_EX_LAYERED | WS_EX_TRANSPARENT));
}
public void SetFormNormal() =>
	SetWindowLong(Handle, GWL_EXSTYLE, Convert.ToInt32(PreviousStyle | WS_EX_LAYERED));
```

flamewave000 commented on DirectXHost.cs:
```diff
@@ -241,134 +272,159 @@ private void _dxForm_HelpRequested(object sender, EventArgs eventArgs)
+                        ShouldShowOverlayFrame(true);
+                    }
+                }),
+                new MenuItem($"{(Settings.frameRate > 0?Settings.frameRate.ToString():"Unlimited")} FPS", (menuItem, e) => {
```
NIT: The ternary operator should have spacing around it for better legibility and the round brackets are unnecessary
```C#
$"{Settings.frameRate > 0 ? Settings.frameRate.ToString() : "Unlimited"} FPS"
```

flamewave000 commented on Settings.cs:
```diff
@@ -52,6 +54,18 @@ public static int frameRate
+			set => data.hostTransparency = value;
+		}
+		
+		public static bool hostTransparent
+		{
+			get => data.hostTransparent;
+			set => data.hostTransparent = value;
+		}
 
```
This value can be calculated and does not need to be saved. NIT: Should be prefixed with "is" to avoid confusion with the `hostTransparency` property

```C#
public static bool hostTransparent => data.hostTransparency != 1;
```

nznaza replied:
using `isHostTransparent` and `hostOpacity` which are more clear

flamewave000 commented on DirectXHost.cs:
```diff
@@ -8,230 +8,261 @@
+            _overlayForm.Controls.Add(_overlayTarget);
+            _overlayForm.ResizeEnd += (sender, args) => { Settings.overlayRect.Size = _overlayForm.ClientSize; Settings.Save(); };
+            _overlayForm.LocationChanged += (sender, args) => { Settings.overlayRect.Point = _overlayForm.Location; Settings.Save(); };
+            _overlayForm.Width = _dxForm.Width; _overlayForm.Height = _dxForm.Height - SystemInformation.MenuHeight;
```
NIT: These should be on separate lines

nznaza replied:
overlooked this one

flamewave000 commented on DirectXHost.cs:
```diff
@@ -241,134 +272,159 @@ private void _dxForm_HelpRequested(object sender, EventArgs eventArgs)
+                        case 0: Settings.frameRate = 5; break;
+                    }
+                    Settings.Save();
+                    (menuItem as MenuItem).Text = $"{(Settings.frameRate > 0?Settings.frameRate.ToString():"Unlimited")} FPS";
```
Same as previous
```C#
(menuItem as MenuItem).Text = $"{Settings.frameRate > 0 ? Settings.frameRate.ToString() : "Unlimited"} FPS";
```

ShowMeBillyJo commented:
@flamewave000 @nznaza May I enquire as to the status of this PR? I've encountered some of the bugs it fixes, and it'd be great to see them brought into the project proper. Thanks!

nznaza commented:
I submitted changes for review addressing all the issues back then, at least as far as I can remember, I can adress any more issues from this PR if present.

flamewave000 commented:
Thanks @nznaza for these changes. I honestly completely forgot about this PR. I'm really sorry about that, I don't really use this program much right now because I've started running my D&D games in person instead of on FoundryVTT/Roll20 so I've lost some focus. I'll get a release made with these changes though.
